File: Others/PG_소수찾기.js
// 순열 알고리즘 연습

function solution(numbers) {
  var answer = 0;
  const num = 1000000
  let chk = new Array(num).fill(1)
  chk[0] = chk[1] = 0
  let idx = 2
  while (idx < num) {
      if (chk[idx] === 1) {
          for (let i = idx*2; i < num; i += idx) {
              chk[i] = 0
          }
      }
      idx++
  }
  let powerSetCheck = new Array(numbers.length).fill(0);
  let powerSetArr = []
  const solve = (depth) => {
      if (depth === powerSetCheck.length) {
          let temp = ''
          for (let i = 0; i < numbers.length; i++) {
              if (powerSetCheck[i] === 1) {
                  temp += numbers[i]
              }
          }
          powerSetArr.push(temp)
      } else {
          //포함되는 경우
          powerSetCheck[depth] = 1
          solve(depth + 1)
          //포함되지 않는 경우
          powerSetCheck[depth] = 0
          solve(depth + 1)
      }
  }
  solve(0)
  console.log(powerSetArr)
  // for (let arr of powerSetArr) {
  //     if (arr.length > 0) {
  //         let prime = parseInt(arr.reduce((a, b) => a+b))
  //         console.log(prime, chk[prime])
  //         if (chk[prime] === 1) {
  //             answer++
  //             chk[prime] = 2
  //             console.log(prime, chk[prime])
  //         }
  //     }        
  // }
  console.log(parseInt("011"))
  console.log(parseInt(numbers[0]))
  
  return answer;
}

File: Others/PG_소수찾기.py
# 프로그래머스 소수찾기
## 부분집합과 순열 문제

from itertools import permutations

def getNum(a):
    result = [1] * (a + 1)
    result[0] = 0
    result[1] = 0
    k = 2
    while k <= a:
        if result[k]:
            for i in range(k * 2, a + 1, k):
                result[i] = 0
        k += 1
    return result


def solution(numbers):
    answer = 0
    nums = [n for n in numbers]
    nums.sort(reverse=True)
    temp_max = ''
    for n in nums:
        temp_max += n
    max_num = int(temp_max)
    chk = getNum(max_num)

    for i in range(1, len(numbers) + 1):
        orders = list(permutations(nums, i))
        for order in orders:
            temp = ''
            for o in order:
                temp += o
            if chk[int(temp)] > 0:
                answer += 1
                chk[int(temp)] = -1

    return answer